File: python/VCFOneParentImputed.py
from __future__ import annotations

# coding: utf-8
# VCFOneParentImputed.py
# 片親だけimputeされていてもう片親はknown

from VCF import VCFSmall
from GenoRecord import GenoRecord
from VCFFamily import VCFFamilyRecord
from VCFImputable import VCFImputable
from ParentProgenyImputer import ParentProgenyImputer
from Map import *


#################### VCFOneParentImputed ####################

class VCFOneParentImputed(VCFImputable):
	def __init__(self, samples: list[str], records: list[VCFFamilyRecord],
					ref_haps: list[list[int]], should_impute_mat: bool,
					map_: Map, vcf: VCFSmall):
		VCFImputable.__init__(self, samples, vcf)
		self.records: list[VCFFamilyRecord] = records
		self.imputer = ParentProgenyImputer(records, ref_haps,
												should_impute_mat, map_, 0.01)
		self.should_impute_mat = should_impute_mat
	
	##### virtual methods for VCFGenoBase #####
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_records(self) -> list[GenoRecord]:
		return [ r for r in self.records ]
	
	##### virtual methods for VCFFamilyBase #####
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	##### virtual methods for VCFImputable #####
	def impute(self) -> None:
		self.imputer.impute()

File: src/VCFOneParentImputed.cpp
#include <cmath>
#include "../include/VCFOneParentImputed.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFOneParentImputed ////////////////////

VCFOneParentImputed::VCFOneParentImputed(const STRVEC& s,
									const std::vector<VCFFamilyRecord *>& rs,
									const std::vector<std::vector<int>>& ref_hs,
									bool should_impute_mat_, const Map& map_,
									double w, const VCFSmall *vcf) :
						VCFImputable(s, vcf),
						records(rs),
						ref_haps(ref_hs),
						imputer(records, should_impute_mat_, ref_hs, map_, w),
						should_impute_mat(should_impute_mat_) { }

VCFOneParentImputed::~VCFOneParentImputed() {
	Common::delete_all(records);
}

size_t VCFOneParentImputed::amount() const {
	const size_t	NH = ref_haps.size();
	const size_t	N = samples.size() - 2;
	return NH*NH * (NH+2*N-1) << (2*N);
}

STRVEC VCFOneParentImputed::imputed_samples() const {
	STRVEC	ss(samples.size()-1);
	const size_t	index = should_impute_mat ? 0 : 1;
	ss[0] = samples[index];
	std::copy(samples.begin() + 2, samples.end(), ss.begin() + 1);
	return ss;
}

void VCFOneParentImputed::impute() {
	imputer.impute();
}
